src/rgw/driver/rados/topic.cc

```diff
@@ -0,0 +1,383 @@
+#include "rgw_common.h"
+#include "rgw_metadata.h"
+#include "rgw_metadata_lister.h"
+#include "rgw_pubsub.h"
+#include "rgw_rados.h"
+#include "rgw_string.h"
+#include "rgw_tools.h"
+#include "rgw_zone.h"
+#include "svc_mdlog.h"
+#include "svc_sys_obj_cache.h"
+
+namespace rgwrados::topic {
+
+static const std::string oid_prefix = "topic.";
+static constexpr std::string_view buckets_oid_prefix = "buckets.";
+
+static rgw_raw_obj get_topic_obj(const RGWZoneParams& zone,
+                                 std::string_view metadata_key)
+{
+  std::string oid = string_cat_reserve(oid_prefix, metadata_key);
+  return {zone.topics_pool, std::move(oid)};
+}
+
+static rgw_raw_obj get_buckets_obj(const RGWZoneParams& zone,
+                                   std::string_view metadata_key)
+{
+  std::string oid = string_cat_reserve(buckets_oid_prefix, metadata_key);
+  return {zone.topics_pool, std::move(oid)};
+}
+
+
+int read(const DoutPrefixProvider* dpp, optional_yield y,
+         RGWSI_SysObj& sysobj, RGWSI_SysObj_Cache& cache_svc,
+         const RGWZoneParams& zone, const std::string& topic_key,
+         rgw_pubsub_topic& info, RGWChainedCacheImpl<cache_entry>& cache,
+         ceph::real_time* pmtime, RGWObjVersionTracker* pobjv)
+{
+  if (auto e = cache.find(topic_key)) {
+    if (pmtime) {
+      *pmtime = e->mtime;
+    }
+    if (pobjv) {
+      *pobjv = std::move(e->objv);
+    }
+    info = std::move(e->info);
+    return 0;
+  }
+
+  const rgw_raw_obj obj = get_topic_obj(zone, topic_key);
+
+  bufferlist bl;
+  cache_entry entry;
+  rgw_cache_entry_info cache_info;
+  int r = rgw_get_system_obj(&sysobj, obj.pool, obj.oid, bl, &entry.objv,
+                             &entry.mtime, y, dpp, nullptr, &cache_info);
+  if (r < 0) {
+    return r;
+  }
+
+  try {
+    auto p = bl.cbegin();
+    decode(entry.info, p);
+  } catch (const buffer::error&) {
+    return -EIO;
+  }
+
+  cache.put(dpp, &cache_svc, topic_key, &entry, {&cache_info});
+
+  if (pmtime) {
+    *pmtime = entry.mtime;
+  }
+  if (pobjv) {
+    *pobjv = std::move(entry.objv);
+  }
+  info = std::move(entry.info);
+  return 0;
+}
+
+int write(const DoutPrefixProvider* dpp, optional_yield y,
+          RGWSI_SysObj& sysobj, RGWSI_MDLog* mdlog, const RGWZoneParams& zone,
+          const rgw_pubsub_topic& info, RGWObjVersionTracker& objv,
+          ceph::real_time mtime, bool exclusive)
+{
+  const std::string topic_key = get_topic_metadata_key(info.user.tenant, info.name);
+  const rgw_raw_obj obj = get_topic_obj(zone, topic_key);
+
+  bufferlist bl;
+  encode(info, bl);
+
+  int r = rgw_put_system_obj(dpp, &sysobj, obj.pool, obj.oid,
+                             bl, exclusive, &objv, mtime, y);
+  if (r < 0) {
+    ldpp_dout(dpp, 1) << "ERROR: failed to write topic obj " << obj.oid
+        << " with: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  // record in the mdlog on success
+  if (mdlog) {
+    return mdlog->complete_entry(dpp, y, "topic", topic_key, &objv);
```

Comment:
So in earlier implementation, the mdlog status was different for write and remove (`MDLOG_STATUS_WRITE` n `MDLOG_STATUS_REMOVE`) and then accordingly the `RGWMetaSyncSingleEntryCR` or `RGWMetaRemoveEntryCR` was invoked

and in current code, both cases mdlog_status is `MDLOG_STATUS_COMPLETE`.
so while syncing how will the secondary zone know whether to delete the topic or write the topic ?